### PR TITLE
DPMMA-2506 Smart Campaign Event Scheduling Based on Contact Behavior

### PR DESCRIPTION
**Pull request by patrykgruszka**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ N ]
| New feature/enhancement? (use the a.x branch)      | [ Y ]
| Deprecations?                          | [ N ]
| BC breaks? (use the c.x branch)        | [ N ]
| Automated tests included?              | [ Y ] 
| Related user documentation PR URL      | https://github.com/mautic/user-documentation/pull/302
| Related developer documentation PR URL | ---
| Issue(s) addressed                     | ---

#### Description:
This pull request enhances campaign system by dynamically optimizing the timing of event execution based on
individual contact behaviors. By analyzing interaction patterns, such as email opens, website visits and form submissions, the system determines the most effective moment to send out marketing communications, thereby increasing the likelihood of
engagement.

This event schedule option is available for Send Email, Marketing Message, Push Contact to Integration and Send a
Webhook

image

#### How It Works

1. **Interaction Data Retrieval**: The system retrieves interaction data for the contact, including email reads, page
   hits and form submissions, to analyze the contact's engagement patterns.
2. **Minimum Interactions Requirement**: A contact must have a minimum number of interactions for the system to
   calculate optimal timing. Each interaction type (e.g., email open, website visit) is only counted once per hour. A
   minimum of 5 interactions is required, otherwise the default ranges are used.
3. **Optimal Time Calculation (execute event within 24 hours)**: Based on the contact's interaction data, the system
   calculates the optimal time for executing a campaign event, considering the contact's historical engagement patterns.
    - If the contact is at that moment within the optimal window, the event will be executed immediately.
    - If the current time is before today's optimal window, the event is scheduled for the first hour of that window.
    - If the current time is after today's optimal window, the event is scheduled for the first hour of the next day's
      optimal window.
4. **Optimal Date Calculation (execute event within 7 days)**: The system calculates the optimal time as in the previous
   option and additionally determines the best day of the week. By default, a contact can have from 1 to 3 optimal days.
    - If the contact is at that moment within the optimal window, the event will be executed immediately.
    - If the current time and day are before the optimal window, then schedule it for the first optimal hour of the next
      optimal day.

#### Default Settings for Optimal Timing

When a contact does not have enough interaction data, default settings are used. These can be configured by an
administrator in the Mautic configuration file and are adjusted according to the contact's preferred timezone.

- **Default Optimal Hour Range**: 9 AM to 12 PM
- **Default Optimal Days**: Tuesday, Wednesday, and Thursday

```
'peak_interaction_timer_best_default_hour_start' => 9,  // 9 AM
'peak_interaction_timer_best_default_hour_end'   => 12, // 12 PM
'peak_interaction_timer_best_default_days'       => [2, 3, 4], // Tuesday, Wednesday, Thursday
```
Based on https://blog.hubspot.com/marketing/best-time-to-send-email

#### Cache and Limits on Interaction Data Retrieval

To optimize performance for large databases, caching is used to store interaction data, reducing the need for frequent
retrieval. The default settings are:

```plaintext
// Keep the cache for 1 month
'peak_interaction_timer_cache_timeout' => 43800,

// Pull only interactions for the last 60 days
'peak_interaction_timer_fetch_interactions_from' => '-60 days',

// Do not pull more than 50 interactions of each type (page hits / email reads)
'peak_interaction_timer_fetch_limit' => 50,
```

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Run the migration
3. Add contacts from various timezones with no interaction history to a segment.
4. Set up a campaign with an email send event, scheduling it for the contact's optimal timing. (Add variant to execute
   event within 24h and a 7d period)
5. Make some interactions in different times and days than the default. Remember, only one interaction of each type (
   email read and page hit) is counted per hour. By default, 5 interactions are needed for the contact to change the
   execution time.
6. Create a new campaign with the same event and check if the event execution times are adjusted to the contact's
   activity time.

### CODE REVIEW

File: app/bundles/CampaignBundle/Executioner/Scheduler/Mode/Optimized.php
Comment (tomekkowalczyk) [inline app/bundles/CampaignBundle/Executioner/Scheduler/Mode/Optimized.php:2]: use `declare(strict_types=1)`

File: app/bundles/LeadBundle/Services/PeakInteractionTimer.php
Comment (tomekkowalczyk) [inline app/bundles/LeadBundle/Services/PeakInteractionTimer.php:2]: use `declare(strict_types=1)`

File: app/bundles/LeadBundle/Config/config.php
Comment (tomekkowalczyk) [inline app/bundles/LeadBundle/Config/config.php]: I suggest providing a unit of time in the comment and giving e.g. hours as the basic unit

File: app/bundles/LeadBundle/Services/PeakInteractionTimer.php
Comment (tomekkowalczyk) [inline app/bundles/LeadBundle/Services/PeakInteractionTimer.php:297]: we can add `@throws \Exception`

File: app/bundles/LeadBundle/Tests/Services/PeakInteractionTimerTest.php
Comment (tomekkowalczyk) [inline app/bundles/LeadBundle/Tests/Services/PeakInteractionTimerTest.php:2]: `declare(strict_types=1);`

**Review by tomekkowalczyk — COMMENTED** [review]
@patrykgruszka very good job, extensive tests make analysis easier. Topics that can be addressed
1. Do we include e-mail opens and website visits, or do we also include form results? I think this is also an important interaction in the system.
2. Can you add a material based on which the default email sending time values are set?
3. We could in the future move variables from `app/bundles/LeadBundle/Config/config.php` to the dashboard configuration
4. A nice functionality would be to add information about the optimal interaction time of a given user on the contact card. I would see this as a CRON job. We could show data from the cache and write the date of last modification next to it.

**Comment by kuzmany** [comment]
We also have a feature like that for segmenting emails.
Have you tested its performance with millions of email_stats and page_hit records?
I will look into it in the coming days.
Thank you.

**Comment by kuzmany** [comment]
@patrykgruszka do you use it already in production?

**Comment by patrykgruszka** [comment]
@tomekkowalczyk good idea, I've added the form submissions to the analysed interactions.

@kuzmany I've tested the queries on a large database and it looks good:
```
0.049 sec in page_hits length > 7 mln
0.019 sec in email_stats length > 18 mln
0.011 sec in form_submissions length > 300k
```
In summary, it's less than 0.1 sec for a single contact to get the data from the DB. Processing the interactions with PHP will not be time consuming, so we can omit that. Additionally, a cache is applied so that the database data is not loaded every time.
We are not using this on production yet. This functionality can be easily added in the emails to segments.

**Comment by kuzmany** [comment]
@patrykgruszka We've already done it for segment emails.
Thank you I will review it this week.

**Review by tomekkowalczyk — APPROVED** [review]
The changes look ok to me, good job. I tested several scenarios and it seems to work according to the task description.

**Comment by RCheesley** [comment]
@patrykgruszka could you take a look at the conflicts on this PR? We will also need some documentation to both add the info about the feature in the email sending section, and the configuration settings for the config page.  Looks like an awesome feature and would love to see it in Mautic!

**Comment by patrykgruszka** [comment]
@RCheesley sure, I've updated this PR and created the user documentation.

**Comment by RCheesley** [comment]
I've tried to test this but despite my best efforts and ensuring that the contact has more than 5 interactions (email opens, page reads, form submissions) and none which are within the same hour, I'm still getting the default time of 9am being set for the email schedule time.

screenshot-mautic ddev site-2024 05 07-15_57_11

Any chance of providing for example an sql export we can run to ensure we have the right data set?

**Comment by patrykgruszka** [comment]
@RCheesley I think in your case the problem may be the fixture data, I'm not sure what it looks like in your database but I suspect that you have somehow cloned it to populate the interaction data. 

I've found a small bug with the minimum interaction condition: greater than -> greater than or equal. However, it should not cause any problems in your case because you have many more interactions than the default 5.

It's best to create the interactions like the real contact would, because then we're sure it will work. But I'll give you the instruction to fill in the page hits anyway, because it's just easier to test.

1. Run command: `ddev ssh` (`ddev exec` with this query was not working well with this example). 
2. Find the contact ID, and replace the number 210 it in the next command  and execute it:
```
php bin/console doctrine:query:sql "INSERT INTO db.page_hits (page_id, redirect_id, email_id, lead_id, ip_id, device_id, date_hit, date_left, country,
    region, city, isp, organization, code, referer, url, url_title, user_agent, remote_host,
    page_language, browser_languages, tracking_id, source, source_id, query)
VALUES (null, null, null, 210, null, null, '2024-05-12 13:51:55', null, '', '', '', '', '', 200, 'https://example.com/',
    'https://example.com/', null, 'PostmanRuntime/7.29.4', null, null,
    '', 'yqst4n7kqy1slxf148zw3sz', null, null,
    ''),
    (null, null, null, 210, null, null, '2024-05-12 14:51:55', null, '', '', '', '', '', 200, 'https://example.com/',
    'https://example.com/', null, 'PostmanRuntime/7.29.4', null, null,
    '', 'yqst4n7kqy1slxf148zw3sz', null, null,
    ''),
    (null, null, null, 210, null, null, '2024-05-12 16:51:55', null, '', '', '', '', '', 200, 'https://example.com/',
    'https://example.com/', null, 'PostmanRuntime/7.29.4', null, null,
    '', 'yqst4n7kqy1slxf148zw3sz', null, null,
    ''),
    (null, null, null, 210, null, null, '2024-05-12 17:51:55', null, '', '', '', '', '', 200, 'https://example.com/',
    'https://example.com/', null, 'PostmanRuntime/7.29.4', null, null,
    '', 'yqst4n7kqy1slxf148zw3sz', null, null,
    ''),
    (null, null, null, 210, null, null, '2024-05-13 15:51:55', null, '', '', '', '', '', 200, 'https://example.com/',
    'https://example.com/', null, 'PostmanRuntime/7.29.4', null, null,
    '', 'yqst4n7kqy1slxf148zw3sz', null, null,
    ''),
    (null, null, null, 210, null, null, '2024-05-13 16:12:55', null, '', '', '', '', '', 200, 'https://example.com/',
    'https://example.com/', null, 'PostmanRuntime/7.29.4', null, null,
    '', 'yqst4n7kqy1slxf148zw3sz', null, null,
    '');"
```

**Comment by RCheesley** [comment]
Thanks @patrykgruszka actually I did make all of the interactions myself by sending the emails and visiting the pages. I then changed the time in the database to ensure they met the conditions that you stipulated. I'll give it another go with the data you suggested!

**Comment by patrykgruszka** [comment]
@RCheesley interesting, if you still have a copy of the database you can send it to me so I can analyse the problem.

**Review by kuzmany — DISMISSED** [review]
I did basic tests, with and without interactions of contact and it worked as expected.

**Comment by kuzmany** [comment]
@patrykgruszka can you resolve conflicts?

**Comment by patrykgruszka** [comment]
Thanks for notification. Conflicts resolved.

**Comment by escopecz** [comment]
@RCheesley would you mind to re-test or provide the database dump this PR? I don't want to merge this when you found issues.

**Comment by patrykgruszka** [comment]
I have verified Ruth's database, which I received in May. I was unable to reproduce the problems. It appears to be a environment/cache issue.

Here is my note:

Calculated optimal window for contact ThomasJDomingue@armyspy.com is between 12:00 and 14:00, looks fine
Calculated optimal days for this contact: Mon, Tue, Sun
image
image

Nevertheless, it would be advisable to test it again before merging, as sending emails is an important part of Mautic.

**Review by kuzmany — APPROVED** [review]
I did basic tests again and it works. 
Also code seems legit.
Good to go 👍

**Comment by escopecz** [comment]
@patrykgruszka it seems this PR caused a regression bug. Can you please take a look at the discussion in https://github.com/mautic/mautic/pull/14325?

**Pull request merged**
